### ISSUE TITLE
[layernorm] Two device hangs with ROW_MAJOR gamma/beta on the non-distributed interleaved path

### ISSUE THREAD

**Issue by kinginu** [tenstorrent/tt-metal#49975]
### Component / Area

ops / normalization (`ttnn.layer_norm` / `ttnn.rms_norm`, interleaved)

### Issue Type

Deterministic Hang

### Observed

On the non-distributed interleaved path, ROW_MAJOR gamma/beta hangs the device in two distinct configurations. The host blocks forever in `ttnn.from_device(out)` while a device core spins at 100%; the board is left wedged and needs a cold power cycle afterwards.

Both were found while adding a `gamma_layout` axis to `test_layernorm_interleaved_all_config` for #49902 (PR #49974) — the fp32 × ROW_MAJOR gamma cell that no test had ever run. They are **separate from #49902**, which is a stick-read byte-count bug causing silent corruption on the legacy path.

**(A) ROW_MAJOR input + ROW_MAJOR gamma/beta — hangs for any input dtype**

| input dtype | gamma dtype | result |
|---|---|---|
| fp32 | bf16 | **hang** |
| bf16 | bf16 | **hang** |

The input dtype is not the trigger here; the ROW_MAJOR input + ROW_MAJOR gamma combination is. Note this kills the plainest configuration there is — bf16 input with bf16 gamma — so it is not a mixed-precision edge case.

**(B) Welford + fp32 input + ROW_MAJOR gamma/beta (TILE input) — the fp32 input is the trigger**

| input dtype | gamma bf16 | gamma fp32 |
|---|---|---|
| **fp32** | **hang** | **hang** |
| bf16 | ✅ 0.999992 | ✅ 0.999991 |
| bf8 | ✅ 0.999927 | ✅ 0.999929 |

The gamma dtype is irrelevant — only the fp32 *input* hangs. `Welford + fp32 input + TILE gamma` passes, so ROW_MAJOR gamma is required.

For contrast, every other ROW_MAJOR-gamma configuration passes:

| welford | input layout | input dtype | result |
|---|---|---|---|
| legacy | TILE | fp32 | ✅ 0.999975 |
| legacy | TILE | bf16 | ✅ 0.999976 |
| legacy | TILE | bf8 | ✅ 0.999922 |
| welford | TILE | bf16 | ✅ 0.999991 |
| welford | TILE | bf8 | ✅ 0.999929 |

py-spy on a hung process:

```
Thread (idle): "MainThread"
    __call__ (ttnn/decorators.py:650)
    test_layernorm_interleaved_all_config (test_layernorm.py:431)   # ttnn.from_device(out)
```

The Python main thread is blocked on readback while the process burns 100% CPU in a device-polling thread — the op never completes on device. The stack is unchanged after 30s and the run never terminates (killed at 41 min in one case).

This is the same class of bug as #43913 (same reader, `reader_unary_interleaved_ln_rm_gb.cpp`): the compute kernel waiting on a CB the reader never pushes. Neither ROW_MAJOR gamma × ROW_MAJOR input nor ROW_MAJOR gamma × Welford has ever been exercised by a test.

### Expected

Either produce correct results, or reject the combinations with a `TT_FATAL`. A device hang requiring a cold power cycle is the worst outcome.

### 1. Steps (exact commands)

```bash
cd $TT_METAL_HOME && source python_env/bin/activate

# (A) ROW_MAJOR input + ROW_MAJOR gamma
python probe_ln_hang.py 0 rm fp32 bf16 rm    # hangs
python probe_ln_hang.py 0 rm bf16 bf16 rm    # hangs

# (B) Welford + fp32 input + ROW_MAJOR gamma
python probe_ln_hang.py 1 rm fp32 fp32       # hangs
python probe_ln_hang.py 1 rm bf16 fp32       # passes, PCC 0.999991
python probe_ln_hang.py 0 rm fp32 fp32       # legacy + TILE input: passes
```

**Each hang wedges the board**, so run one case per boot — results measured after a previous hang are not trustworthy.

<details><summary>probe_ln_hang.py (self-contained)</summary>

```python
"""Usage: probe_ln_hang.py <welford:0|1> <gamma_layout:tile|rm> <in_dtype> <gamma_dtype> [in_layout:tile|rm]"""
import sys
import torch
import ttnn

M, K = 256, 1024
dtypes = {"fp32": ttnn.float32, "bf16": ttnn.bfloat16, "bf8": ttnn.bfloat8_b}
welford = sys.argv[1] == "1"
gl = ttnn.ROW_MAJOR_LAYOUT if sys.argv[2] == "rm" else ttnn.TILE_LAYOUT
dt, gdt = dtypes[sys.argv[3]], dtypes[sys.argv[4]]
il = ttnn.ROW_MAJOR_LAYOUT if (len(sys.argv) > 5 and sys.argv[5] == "rm") else ttnn.TILE_LAYOUT

device = ttnn.open_device(device_id=0)
try:
    torch.manual_seed(0)
    x = torch.rand((1, 1, M, K), dtype=torch.float32)
    w = torch.rand((K,), dtype=torch.float32)
    b = torch.rand((K,), dtype=torch.float32)
    ref = torch.nn.functional.layer_norm(x, (K,), weight=w, bias=b, eps=1e-12)

    xt = ttnn.from_torch(x, dtype=dt, layout=il, device=device, memory_config=ttnn.DRAM_MEMORY_CONFIG)
    gb_shape = (1, 1, K // 32, 32) if gl == ttnn.ROW_MAJOR_LAYOUT else (1, 1, 1, K)
    wt = ttnn.from_torch(w.reshape(gb_shape), dtype=gdt, layout=gl, device=device)
    bt = ttnn.from_torch(b.reshape(gb_shape), dtype=gdt, layout=gl, device=device)

    ckc = ttnn.init_device_compute_kernel_config(
        device.arch(), math_fidelity=ttnn.MathFidelity.HiFi4, math_approx_mode=False,
        fp32_dest_acc_en=True, packer_l1_acc=False,
    )
    cfg = ttnn.LayerNormDefaultProgramConfig(use_welford=welford)

    recip = None
    if welford:
        grid = device.compute_with_storage_grid_size()
        crs = ttnn.CoreRangeSet({ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(grid.x - 1, grid.y - 1))})
        recip = ttnn.create_layer_norm_reciprocals(device, crs, K)

    print("dispatching...", flush=True)
    out = ttnn.layer_norm(xt, epsilon=1e-12, weight=wt, bias=bt, program_config=cfg,
                          compute_kernel_config=ckc, recip_tensor=recip)
    ot = ttnn.to_torch(ttnn.from_device(out)).float().reshape(ref.shape)   # <-- hangs here
    a, c = ref.flatten().float(), ot.flatten().float()
    print(f"OK PCC {torch.corrcoef(torch.stack([a, c]))[0, 1].item():.6f}", flush=True)
finally:
    ttnn.close_device(device)
```
</details>

### 2. Input data / link or description

None. Self-contained — `torch.rand` only.

### 3. Frequency

Always (100%), deterministic.

### 1. Software Versions

```
OS:               Ubuntu 24.04.3 LTS (Noble Numbat)
tt-metal:         upstream/main @ 1cb6f8e40e (2026-07-15)
Python:           3.12.12
torch:            2.11.0+cpu
KMD:              2.7.0
Firmware bundle:  19.7.0
```

### 2. Hardware Details

```
Product:      Blackhole
Card/System:  p300c x4 (QuietBox); repro uses device 0 only
```

### Is this a regression?

No — neither combination has ever been covered by a test.

### Regression Details

Not a regression. `test_rms_norm_row_major` (the only pre-existing ROW_MAJOR-gamma test) is bf16-only, TILE-input, and non-Welford. `test_layernorm_interleaved_all_config` (added by #48226) sweeps input layout and gamma dtype but builds gamma as TILE. Neither ROW_MAJOR gamma × ROW_MAJOR input nor ROW_MAJOR gamma × Welford has ever run.

### Logs & Diagnostics

```
# after the hang is killed, the board stays wedged:
TT_THROW: Device 0: Timeout (10000 ms) waiting for physical cores to finish: 4-2, 2-2, 10-2, 7-2, 6-2, 1-2, 3-2, 5-2.
TT_THROW: Device 0 init: failed to initialize FW! Try resetting the board.
```

### Impact

Any model using ROW_MAJOR norm weights with either ROW_MAJOR input or Welford + fp32 input hangs the board. Found alongside #49902 during Qwen3.5 bring-up on Blackhole. **Workaround:** use TILE-layout gamma/beta, or TILE input with the legacy (non-Welford) path.

### Notes — provenance of the measurements

Verified pre-existing, i.e. **not** caused by #49974 (the #49902 fix):

- (A) with fp32 input and (B) with fp32 gamma were reproduced directly on **pristine `upstream/main`**, with no patches applied.
- The remaining hang measurements used bf16 gamma. For bf16 gamma, #49974 computes `row_bytes=64, half_row_bytes=32, face_bytes=512` — **byte-for-byte identical to the hardcoded values it replaces** — so the kernel is behaviourally unchanged for those cases, and the hangs are pre-existing there too.

PR #49974 skips both combinations in `test_layernorm_interleaved_all_config` and references this issue, so the matrix records the limitations rather than hanging CI — mirroring the existing `welford × rm_in` skip in the same test.

I have a root cause for both and am working on a fix PR.
